File: packages/core/src/visibility.ts
import type { Visibility } from "./types.js";

/** Rank: private=0, unlisted=1, public=2 — min rank = most strict (KD-21). */
const RANK: Record<Visibility, number> = {
  private: 0,
  unlisted: 1,
  public: 2,
};

const FROM_RANK: Visibility[] = ["private", "unlisted", "public"];

/** Coerce unknown stored values to a valid Visibility (F-002). */
export function asVisibility(v: unknown, fallback: Visibility = "private"): Visibility {
  if (v === "private" || v === "unlisted" || v === "public") return v;
  return fallback;
}

export function visibilityRank(v: Visibility | string): number {
  return RANK[asVisibility(v)] ?? 0;
}

/**
 * effective_visibility(node) = min_rank(self, ancestors).
 * Pass ancestors from root→parent order or any order — min is commutative.
 * Accepts string[] from storage; narrows safely (F-002).
 */
export function effectiveVisibility(
  self: Visibility | string,
  ancestors: Array<Visibility | string> = [],
): Visibility {
  let rank = visibilityRank(asVisibility(self));
  for (const a of ancestors) {
    rank = Math.min(rank, visibilityRank(asVisibility(a)));
  }
  return FROM_RANK[rank]!;
}

File: server/app/domain/visibility.py
from __future__ import annotations

from collections.abc import Iterable

RANK = {"private": 0, "unlisted": 1, "public": 2}
FROM_RANK = ["private", "unlisted", "public"]


def effective_visibility(self: str, ancestors: Iterable[str] = ()) -> str:
    rank = RANK.get(self, 0)
    for a in ancestors:
        rank = min(rank, RANK.get(a, 0))
    return FROM_RANK[rank]


def is_public_nav_visible(self: str, ancestors: Iterable[str] = ()) -> bool:
    return effective_visibility(self, ancestors) == "public"
